src/esp/gfx/PBRDrawable.cpp

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include "PBRDrawable.h"
+
+#include <Corrade/Containers/ArrayViewStl.h>
+#include <Corrade/Utility/FormatStl.h>
+
+namespace Mn = Magnum;
+
+namespace esp {
+namespace gfx {
+
+PBRDrawable::PBRDrawable(scene::SceneNode& node,
+                         Mn::GL::Mesh& mesh,
+                         ShaderManager& shaderManager,
+                         const Mn::ResourceKey& lightSetupKey,
+                         const Mn::ResourceKey& materialDataKey,
+                         DrawableGroup* group)
+    : Drawable{node, mesh, group},
+      shaderManager_{shaderManager},
+      lightSetup_{shaderManager.get<LightSetup>(lightSetupKey)},
+      materialData_{
+          shaderManager.get<MaterialData, PBRMaterialData>(materialDataKey)} {
+  updateShader().updateShaderLightParameters();
+}
+
+void PBRDrawable::setLightSetup(const Mn::ResourceKey& lightSetup) {
+  lightSetup_ = shaderManager_.get<LightSetup>(lightSetup);
+  updateShader().updateShaderLightParameters();
+}
+
+void PBRDrawable::draw(const Mn::Matrix4& transformationMatrix,
+                       Mn::SceneGraph::Camera3D& camera) {
+  // no need to call updateShaderLightParameters() in the draw loop,
+  // only need to update the position or direction of the lights
+  updateShader().updateShaderLightDirectionParameters(transformationMatrix,
+                                                      camera);
+
+  (*shader_)
+      // e.g., semantic mesh has its own per vertex annotation, which has been
+      // uploaded to GPU so simply pass 0 to the uniform "objectId" in the
+      // fragment shader
+      .setObjectId(
+          static_cast<RenderCamera&>(camera).useDrawableIds()
+              ? drawableId_
+              : (materialData_->perVertexObjectId ? 0 : node_.getSemanticId()))
+      .setTransformationMatrix(transformationMatrix)  // modelview matrix
+      .setMVPMatrix(camera.projectionMatrix() * transformationMatrix)
+      // because we actually do not have scaling, only rotation, so that the
+      // modelview matrix mv = inv(mv^T) where mv stands for the up-left 3x3
+      // part of the modelview matrix
+      .setNormalMatrix(transformationMatrix.rotationScaling())
+      .bindTextures(
+          materialData_->albedoTexture, materialData_->roughnessTexture,
+          materialData_->metallicTexture, materialData_->normalTexture)
+      .setBaseColor(materialData_->baseColor)
+      .setRoughness(materialData_->roughness)
+      .setMetallic(materialData_->metallic);
```

Comment:
Just FYI, in Magnum I'm using "metalness" to be similar to "roughness", but it's a minor thing -- just if you'd like to have the exact same naming scheme here.